keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md

```diff
@@ -0,0 +1,1010 @@
+	skip         bool
+	boundClaims  []*v1.PersistentVolumeClaim
+	claimsToBind []*v1.PersistentVolumeClaim
+	allBound     bool
+	podVolumesByNode map[string]*PodVolumes
+	sync.Mutex
+}
+```
+
+By making the following changes to `PodVolumes`, `CSIStorageCapacity` can be stored.
+
+```diff
++ type DynamicProvision struct {
++ 	PVC      *v1.PersistentVolumeClaim
++ 	Capacity *storagev1.CSIStorageCapacity
++ }
+
+type PodVolumes struct {
+	StaticBindings []*BindingInfo
+-   DynamicProvisions []*v1.PersistentVolumeClaim
++ 	DynamicProvisions []*DynamicProvision
+}
+```
+
+### Get the capacity of nodes for dynamic provisioning
+
+Add `CSIStorageCapacity` to the return value of the `volumeBinder.hasEnoughCapacity` method. This returns the `DynamicProvision.Capacity` field in the case of dynamic provisioning.
+
+```diff
+- func (b *volumeBinder) hasEnoughCapacity(provisioner string, claim *v1.PersistentVolumeClaim, storageClass *storagev1.StorageClass, node *v1.Node) (bool, error) {
++ func (b *volumeBinder) hasEnoughCapacity(provisioner string, claim *v1.PersistentVolumeClaim, storageClass *storagev1.StorageClass, node *v1.Node) (bool, *storagev1.CSIStorageCapacity, error) {
+	quantity, ok := claim.Spec.Resources.Requests[v1.ResourceStorage]
+	if !ok {
+		// No capacity to check for.
+- 		return true, nil
++ 		return true, nil, nil
+	}
+
+	// Only enabled for CSI drivers which opt into it.
+	driver, err := b.csiDriverLister.Get(provisioner)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Either the provisioner is not a CSI driver or the driver does not
+			// opt into storage capacity scheduling. Either way, skip
+			// capacity checking.
+- 			return true, nil
++ 			return true, nil, nil
+		}
+- 		return false, err
++ 		return false, nil, err
+	}
+	if driver.Spec.StorageCapacity == nil || !*driver.Spec.StorageCapacity {
+- 		return true, nil
++ 		return true, nil, nil
+	}
+
+	// Look for a matching CSIStorageCapacity object(s).
+	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure (https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718).
+	capacities, err := b.csiStorageCapacityLister.List(labels.Everything())
+	if err != nil {
+- 		return false, err
++ 		return false, nil, err
+	}
+
+  sizeInBytes := quantity.Value()
+	for _, capacity := range capacities {
+		if capacity.StorageClassName == storageClass.Name &&
+			capacitySufficient(capacity, sizeInBytes) &&
+			b.nodeHasAccess(node, capacity) {
+			// Enough capacity found.
+- 			return true, nil
++ 			return true, capacity, nil
+		}
+	}
+
+	// TODO (?): this doesn't give any information about which pools where considered and why
+	// they had to be rejected. Log that above? But that might be a lot of log output...
+	klog.V(4).InfoS("Node has no accessible CSIStorageCapacity with enough capacity for PVC",
+		"node", klog.KObj(node), "PVC", klog.KObj(claim), "size", sizeInBytes, "storageClass", klog.KObj(storageClass))
```

Comment:
That area is in diff format. Since the lines do not have a '+' or '-' at the beginning, those lines already exist. Those are not lines added by this KEP.